### ISSUE TITLE
[PR #1916] fix(dylint): fix lint annotation line reporting

### ISSUE THREAD

**Issue by il10241024** [constructorfabric/gears-rust#3500]
> 🔗 **Mirrored PR** [cyberfabric/cyberware-rust#1916](https://github.com/cyberfabric/cyberware-rust/pull/1916) | **Author:** fluiderson | **Opened:** 2026-05-15T19:15:08Z | **Status:** merged on 2026-05-20T11:14:16Z
> *GitHub API does not allow setting PR author or timestamps — attribution preserved here.*

---

For example,

```rs
// simulated_dir=/cyberfabric/modules/some_module/contract/
use serde::Deserialize;

// Should trigger DE0101 - Serde in contract
#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize)]
pub struct Order {
    pub id: String,
    pub total: f64,
}

#[allow(dead_code)]
// Should trigger DE0101 - Serde in contract
#[derive(Debug, Clone, Deserialize)]
pub enum UserRole {
    Admin,
    User,
    Guest,
}

fn main() {}
```

reports

```
In "deserialize.rs": Line 5 has '// Should trigger DE0101 - Serde in contract' comment but no corresponding error in .stderr file
```

when the expected error is

```
In "deserialize.rs": Line 4 has '// Should trigger DE0101 - Serde in contract' comment but no corresponding error in .stderr file
```

## Summary by CodeRabbit

* **Tests**
  * Enhanced test harness with improved tracking of test expectations and clearer error messages for lint validation failures.

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/cyberfabric/cyberware-rust/pull/1916)

---

**Comment by il10241024** [comment]
**qodo-code-review[bot]** commented on 2026-05-15T19:15:13Z:

---

### Qodo reviews are paused for this user.

Troubleshooting steps vary by plan [Learn more →](https://docs.qodo.ai/review-eligibility)


**On a Teams plan?**
Reviews resume once this user has a paid seat *and* their Git account is linked in Qodo.
[Link Git account →](https://app.qodo.ai)

**Using GitHub Enterprise Server, GitLab Self-Managed, or Bitbucket Data Center?**
These require an Enterprise plan - Contact us
[Contact us →](https://app.qodo.ai)

**Comment by il10241024** [comment]
**coderabbitai[bot]** commented on 2026-05-15T19:15:16Z:

---


<details>
<summary>ℹ️ Recent review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Free

**Run ID**: `e9955cc3-b308-4462-bb4e-0e243d0dece7`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between 5bc1b647a4e5ff7ff95d4abee885d2aa99b10aff and 8d070a6884357bd61ab21598d99090fc4142d59c.

</details>

<details>
<summary>📒 Files selected for processing (1)</summary>

* `tools/dylint_lints/lint_utils/src/lib.rs`

</details>

</details>

---


<details>
<summary>📝 Walkthrough</summary>

## Walkthrough

The PR refactors the test comment annotation helper in the linting utilities to track stderr error lines mapped to their source comment line numbers. It replaces a simple line number set with a HashMap to enable clearer failure messages during test assertion validation.

## Changes

**Test utility comment-to-error mapping**

|Layer / File(s)|Summary|
|---|---|
|**Comment-to-error-line mapping** <br> `tools/dylint_lints/lint_utils/src/lib.rs`|Import change adds `HashMap` and removes `HashSet`. Comment scanning logic stores "should trigger" and "should not trigger" expectations in `HashMap`s keyed by expected stderr error line numbers with originating comment line numbers as values. Assertion loops iterate over the new mapping and reference stored comment line numbers in failure messages for clearer diagnostics.|

## Estimated code review effort

🎯 2 (Simple) | ⏱️ ~10 minutes

## Poem

> 🐰 With maps instead of sets so bare,  
> Test comments now know where they dare!  
> Error lines trace back with care,  
> To source whispers floating in air.  
> No more mystery—clarity's there! ✨

</details>




> [!TIP]
> <details>
> <summary>💬 Introducing Slack Agent: The best way for teams to turn conversations into code.</summary>
> 
> [Slack Agent](https://www.coderabbit.ai/agent) is built on CodeRabbit's deep understanding of your code, so your team can collaborate across the entire SDLC without losing context.
> 
> - Generate code and open pull requests
> - Plan features and break down work
> - Investigate incidents and troubleshoot customer tickets together
> - Automate recurring tasks and respond to alerts with triggers
> - Summarize progress and report instantly
> 
> Built for teams:
> 
> - **Shared memory** across your entire org—no repeating context
> - **Per-thread sandboxes** to safely plan and execute work
> - **Governance built-in**—scoped access, auditability, and budget controls
> 
> One agent for your entire SDLC. Right inside Slack.
> 
> 👉 [Get started](https://agent.coderabbit.ai/)
> 
> </details>




---

> [!NOTE]
> <details>
> <summary>🎁 Summarized by CodeRabbit Free</summary>
> 
> Your organization is on the Free plan. CodeRabbit will generate a high-level summary and a walkthrough for each pull request. For a comprehensive line-by-line review, please upgrade your subscription to CodeRabbit Pro by visiting <https://app.coderabbit.ai/login>.
> 
> </details>


<sub>Comment `@coderabbitai help` to get the list of available commands and usage tips.</sub>
